### PR TITLE
Add vlab discounts

### PR DESCRIPTION
**Pull request by pgetta**
Another small change is addition of an initial balance for vlab account creation.

### CODE REVIEW

File: app/repository/discount.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        Finds the most recent discount that is currently valid (started in the past and
+        either has no end date or ends in the future).
+
+        Args:
+            vlab_id: UUID of the virtual lab to query discounts for
+
+        Returns:
+            The most recent active Discount object or None if no active discount exists
+        """
+        now = datetime.now(tz=UTC)
```
Comment (GianlucaFicarelli) [inline app/repository/discount.py:30]: Minor, there is `utils.utcnow()` that would be easier to mock or change if needed in the future.

File: app/repository/discount.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+        Args:
+            discount_id: The UUID of the discount to update
+            data: Dictionary containing the fields to update (percentage, valid_from, etc.)
+
+        Returns:
+            Updated Discount object or None if discount not found
+        """
+        query = (
+            sa.update(Discount).where(Discount.id == discount_id).values(data).returning(Discount)
```
Comment (GianlucaFicarelli) [inline app/repository/discount.py:89]: I would allow updating a discount only:
1. if `valid_from` is in the future, or
2. if `valid_from` is in the past AND we change only `valid_to` to a date in the future (no changes allowed to `valid_from` and `discount`)

We could also keep only the case 1, because for the use case 2, to change an existing discount already in effect, we can always add another record with valid_from=now.

Or we can completely forbid the update (for the price it's not allowed because of this).

In this way we should still be able to keep track of the history of discounts (but we should also forbid to insert discounts with valid_from in the past, see the other comment).

File: app/schema/api.py [hunk 3]
```diff
@@ -243,6 +245,38 @@ class AddPriceOut(AddPriceIn):
     id: int
 
 
+class AddDiscountIn(BaseModel):
+    """AddDiscountIn."""
+
+    vlab_id: UUID
+    discount: Decimal
+    valid_from: AwareDatetime
+    valid_to: AwareDatetime | None = None
```
Comment (GianlucaFicarelli) [inline app/schema/api.py:254]: For confirmation:
1. do we want to be able to apply a discount at any time during the day? (This is the same for regular prices, there is no enforcement on starting the price at midnight or at any hour during the day)
2.  do we want to allow `valid_from` to be in the past? (This is the same for regular prices, but maybe it's not a good idea, because it could create inconsistencies with the registered entries in the ledger. It could accept None to mean apply it immediately, for example)

These are mainly for discussion, we can open an issue for those.

File: app/schema/api.py [hunk 4]
```diff
@@ -243,6 +245,38 @@ class AddPriceOut(AddPriceIn):
...
+
+    @field_validator("discount")
+    @classmethod
+    def validate_discount(cls, v: Decimal) -> Decimal:
+        """Validate discount."""
+        if v < Decimal(0) or v > Decimal(1):
+            msg = "Discount must be between 0 and 1"
+            raise ValueError(msg)
+        return v
```
Comment (GianlucaFicarelli) [inline app/schema/api.py:263]: This validation is supported by pydantic, something like:
```python
    discount: Annotated[Decimal, Field(gt=D0, lt=Decimal(1))]
```

File: app/api/account.py [hunk 5]
```diff
@@ -34,7 +34,9 @@ async def add_virtual_lab_account(
     repos: RepoGroupDep, account: VlabAccountCreationIn
 ) -> ApiResponse[VlabAccountCreationOut]:
     """Add a new virtual lab account."""
-    result = await account_service.add_virtual_lab(repos, account_id=account.id, name=account.name)
+    result = await account_service.add_virtual_lab(
+        repos, account_id=account.id, name=account.name, balance=account.balance
```
Comment (GianlucaFicarelli) [inline app/api/account.py:38]: Do we want allow to create a virtual-lab with a balance different from 0?
Wouldn't it create a discrepancy with the double entry accounting?

File: app/service/price.py [hunk 6]
```diff
@@ -2,17 +2,27 @@
...
+    discount: Discount | None = None,
+) -> Decimal:
     """Return the cost for a job."""
     cost = price.multiplier * usage_value
     if include_fixed_cost:
         cost += price.fixed_cost
+
+    if discount:
+        cost *= Decimal(1) - discount.discount
+
```
Comment (GianlucaFicarelli) [inline app/service/price.py:25]: Just for confirmation, should the discount be applied to both the variable and the fixed costs?
Better to add an explicit note in that case.

File: app/repository/discount.py (hunk 1 above)
Comment (pgetta) [reply app/repository/discount.py]: Good point, fixed.

File: app/repository/discount.py (hunk 2 above)
Comment (pgetta) [reply app/repository/discount.py]: I've removed the update endpoint/logic, there is no necessity to have it now.

File: app/schema/api.py (hunk 3 above)
Comment (pgetta) [reply app/schema/api.py:254]: 1. yes - any time works, the main use-case for this is when user upgrades an account to a "pro" or "premium" - we want to create a discount right away so that the user can immediately benefit from the subscription.

2. Good point, that doesn't make much sense (at least I can't see use-cases for that). I propose to add extra validation in a separate PR/iteration. Setting none is good idea, indeed.

File: app/schema/api.py (hunk 4 above)
Comment (pgetta) [reply app/schema/api.py]: Implemented, I've added some tests for both ends of ranges as well.

File: app/api/account.py (hunk 5 above)
Comment (pgetta) [reply app/api/account.py:38]: That is extremely good point, Gianluca, thank you for noticing that.
I've updated the service logic to perform an action in two steps: creating an account and after making an top-up transaction on top of that.

File: app/schema/api.py [hunk 7]
```diff
@@ -154,6 +154,7 @@ class VlabAccountCreationIn(BaseModel):
 
     id: UUID
     name: str
+    balance: Decimal = D0
```
Comment (GianlucaFicarelli) [inline app/schema/api.py:157]: We should enforce balance >=0, unless there is a use case to create an account with negative balance.

File: app/schema/api.py [hunk 8]
```diff
@@ -243,6 +245,29 @@ class AddPriceOut(AddPriceIn):
     id: int
 
 
+class AddDiscountIn(BaseModel):
+    """AddDiscountIn."""
+
+    vlab_id: UUID
+    discount: Annotated[Decimal, Field(ge=D0, le=D1)]
```
Comment (GianlucaFicarelli) [inline app/schema/api.py:252]: For clarification, is it acceptable to create a discount with value 0 or 1? My thoughts:
- 0 would be equal to no-discount, so it can be used to invalidate an existing discount since it cannot be updated, for example
- 1 would be equal to make the service free

If these are the intended use cases, we could add a comment somewhere in the code or docs.
We could also test the edge cases explicitly.

File: app/service/account.py [hunk 9]
```diff
@@ -1,19 +1,30 @@
...
+) -> VlabAccount:
     """Add a new virtual lab account."""
-    return await repos.account.add_vlab_account(account_id=account_id, name=name)
+    account = await repos.account.add_vlab_account(account_id=account_id, name=name)
+
+    if balance == D0:
+        return account
+
+    await top_up(repos, vlab_id=account.id, amount=balance)
+    return await repos.account.get_vlab_account(vlab_id=account.id)
```
Comment (GianlucaFicarelli) [inline app/service/account.py:27]: Looks good, but is the use case of creating an account with an initial balance frequent enough to justify the implementation of the logic in the service, instead of asking the client to make 2 calls (create and top-up)?

File: app/schema/api.py (hunk 7 above)
Comment (pgetta) [reply app/schema/api.py]: I've added the validation and a test.

File: app/schema/api.py (hunk 8 above)
Comment (pgetta) [reply app/schema/api.py:252]: Now there is a note under the API docstring, as well as tests for edge cases: for discount 0 and 1.

File: app/service/account.py (hunk 9 above)
Comment (pgetta) [reply app/service/account.py:27]: It's a small optimisation measure and in VLab API depending on the user state it's 50/50: half of the requests will use non-zero initial balance, while the other half will not.

File: app/service/price.py (hunk 6 above)
Comment (pgetta) [reply app/service/price.py:26]: I've added a note.

File: app/schema/api.py [hunk 10]
```diff
@@ -154,7 +154,7 @@ class VlabAccountCreationIn(BaseModel):
 
     id: UUID
     name: str
-    balance: Decimal = D0
+    balance: Annotated[Decimal, Field(default=D0, ge=D0)]
```
Comment (GianlucaFicarelli) [inline app/schema/api.py:157]: Minor, with the Annotated syntax the default can be provided after the equal sign
```suggestion
    balance: Annotated[Decimal, Field(ge=D0)] = D0
```

**Review by GianlucaFicarelli — DISMISSED** [review]

**Review by GianlucaFicarelli — APPROVED** [review]
